test/recovery.js

```diff
@@ -0,0 +1,101 @@
+'use strict'
+/* eslint-env node, mocha */
+
+const TcpConnection = require('../lib/TcpConnection')
+const { expect } = require('chai')
+const net = require('net')
+
+function startServer ({ address, port, next }) {
+  const socket = net.createServer((connection) => {
+    connection.on('data', (data) => {
+      next({ action: 'data', data })
+      connection.destroy()
+    })
+  })
+
+  socket.listen(port || 0, address || '127.0.0.1', () => {
+    next({
+      action: 'started',
+      address: socket.address().address,
+      port: socket.address().port
+    })
+  })
+
+  return socket
+}
+
+test('recovery', function (done) {
+  let address
+  let port
+  let tcpConnection
+  let counter = 0
+  let closing = false
+  const received = []
+
+  function sendData () {
+    setInterval(() => {
+      counter++
+      tcpConnection.write(`log${counter}\n`, 'utf8', () => { /* ignore */ })
+    }, 100)
+  }
+
+  function startSecondServer () {
+    const secondServer = startServer({
+      address,
+      port,
+      next: (msg) => {
+        switch (msg.action) {
+          case 'data':
+            received.push(msg)
+            if (!closing) {
+              closing = true
+              setTimeout(() => {
+                secondServer.close(() => {
+                  try {
+                    const logs = received
+                      .map(it => it.data.toString('utf8'))
+                      .reduce((previousValue, currentValue) => previousValue + currentValue)
+                      .split('\n')
+                      .filter(it => it !== '')
+                    console.log(logs)
+                    const logNumbers = logs.map(it => parseInt(it.replace('log', '')))
+                    expect(logs.length).to.eq(logNumbers[logNumbers.length - 1])
+                    // make sure that no number is missing
+                    expect(logNumbers).to.deep.eq(Array.from({ length: logNumbers.length }, (_, i) => i + 1))
+                  } finally {
+                    done()
+                  }
+                })
+              }, 500) // wait recovery a bit to make sure that enqueued data have been recovered
```

Comment:
Could we use fake timers instead? Timers in tests are a source of test flakyness.

Reply:
I'm now counting the number of events received instead of waiting an arbitrary amount of time